mediatek: improve support for cudy devices

aiamadeus opened the pull request:
Fixes typo for spi and mtd properties.
Fixes interrupt support for 2.5G PHY.
Delete the unused rfb compatible.
Fixes typo for led properties.

aiamadeus committed:
mediatek: cudy: fixes typo for spi and mtd properties

Same as commit 3674689, correct 'buswidth' to 'bus-width'.
Move the nmbm properties outside the partition definition.
Change uppercase to lowercase, add missing read-only flag.

Signed-off-by: Chukun Pan <amadeus@jmu.edu.cn>

aiamadeus committed:
mediatek: update status led for Cudy TR3000

Use white led when running, consistent with the stock firmware.

Signed-off-by: Chukun Pan <amadeus@jmu.edu.cn>

aiamadeus committed:
mediatek: cleanup device tree for cudy devices

Fixes typo for led properties.
Delete the unused rfb compatible.

Signed-off-by: Chukun Pan <amadeus@jmu.edu.cn>

aiamadeus committed:
mediatek: cudy: fixes 2.5G PHY interrupt support

Fixed interrupt support for 2.5G PHY.
Removed useless phy-mode on phy node.

Tested on Cudy TR3000.
Signed-off-by: Chukun Pan <amadeus@jmu.edu.cn>

LuisMitaHL commented:
One question, are the commits for TR3000 just code cleanup? Or do they solve a bug?
There is a discussion that may be relevant: https://github.com/openwrt/openwrt/issues/17183

aiamadeus commented:
> One question, are the commits for TR3000 just code cleanup? Or do they solve a bug?

This does not fix the above problem, I think it is a netifd bug.
I'm busy lately and don't have much time to debug this.

aiamadeus committed:
mediatek: use standard PHY reset bindings for Cudy WR3000H

Use generic Ethernet PHY reset bindings.
Remove the useless lan label as connected switch.

Fixes: 9d66b8b ("mediatek: filogic: Add support for cudy wr3000h")
Signed-off-by: Chukun Pan <amadeus@jmu.edu.cn>

Linaro1985 commented:
@aiamadeus could you add spi section fix (tx bus width and rx bus width) for Cudy AP3000 v1 please?

hauke commented on target/linux/mediatek/dts/mt7981b-cudy-m3000-v1.dts:
```diff
@@ -127,18 +126,18 @@
 		spi-max-frequency = <52000000>;
-		spi-tx-buswidth = <4>;
-		spi-rx-buswidth = <4>;
+		spi-tx-bus-width = <4>;
+		spi-rx-bus-width = <4>;
+
+		mediatek,nmbm;
+		mediatek,bmt-max-ratio = <1>;
+		mediatek,bmt-max-reserved-blocks = <64>;
 
 		partitions {
 			compatible = "fixed-partitions";
 			#address-cells = <1>;
 			#size-cells = <1>;
 
-			mediatek,nmbm;
-			mediatek,bmt-max-ratio = <1>;
-			mediatek,bmt-max-reserved-blocks = <64>;
-
```
Did you or someone else test this change on a device?

Does the device need nmbm?

Linaro1985 replied:
@hauke according original SDK based on OpenWrt the devices needs nmbm.

aiamadeus replied:
The OEM firmware enables nmbm, so we need it.

hauke commented:
On which devices did you test this?

aiamadeus commented:
> On which devices did you test this?

Tested on Cudy TR3000.
It would be nice if someone else tested it.

hauke commented:
@juampe could you please check if the 2.5G PHY of the cudy wr3000h still works after this change.

juampe commented:
Thanks @hauke 

With main works OK with C22
```
root@wrt2:~# dmesg|grep wan
[    2.828857] mtk_soc_eth 15100000.ethernet wan: renamed from eth1
[   13.665401] mtk_soc_eth 15100000.ethernet wan: PHY [mdio-bus:06] driver [RTL8221B-VB-CG 2.5Gbps PHY (C22)] (irq=POLL)
[   13.676055] mtk_soc_eth 15100000.ethernet wan: configuring for phy/2500base-x link mode
[   16.878496] mtk_soc_eth 15100000.ethernet wan: Link is Up - 1Gbps/Full - flow control off
```

With main and 17600 patch:
The Wan interface not work with C45 (remember I tested a lot with this in https://forum.openwrt.org/t/support-for-cudy-wr3000h/217336 )
```
root@wrt2:~# dmesg |grep wan
[    3.136115] mtk_soc_eth 15100000.ethernet wan: renamed from eth1
[   14.676575] mtk_soc_eth 15100000.ethernet wan: PHY [mdio-bus:06] driver [RTL8221B-VB-CG 2.5Gbps PHY (C45)] (irq=POLL)
[   14.687250] mtk_soc_eth 15100000.ethernet wan: configuring for phy/2500base-x link mode
```

No link up, sorry for lack of more test but now is in "production" state.

aiamadeus commented:
Is this similar to this question? #17232

juampe commented:
I don't think is related to https://github.com/openwrt/openwrt/issues/17232
AFAIK with my tests this issue is related to a bad PHY register/reference in dts

```
[   15.556699] RTL8221B-VB-CG 2.5Gbps PHY (C45) mdio-bus:01: rtl822xb_config_init failed: -110
[   15.675872] mtk_soc_eth 15100000.ethernet eth0: mtk_open: could not attach PHY: -110
```

In this proposal the reg is OK and the the reference is ok , but with C45 driver I never got link up , please keep c22

aiamadeus commented:
Can you share the dmesg log without workaroud?

juampe commented:
```
root@wrt2:~# dmesg
[    0.000000] Booting Linux on physical CPU 0x0000000000 [0x410fd034]
[    0.000000] Linux version 6.6.71 (ubuntu@wally) (aarch64-openwrt-linux-musl-gcc (OpenWrt GCC 13.3.0 r28262+7-1269a836ed) 13.3.0, GNU ld (GNU Binutils) 2.42) #0 SMP Tue Jan 21 09:53:03 2025
[    0.000000] Machine model: Cudy WR3000H v1
[    0.000000] OF: reserved mem: 0x0000000042ff0000..0x0000000042ffffff (64 KiB) map non-reusable ramoops@42ff0000
[    0.000000] OF: reserved mem: 0x0000000043000000..0x000000004302ffff (192 KiB) nomap non-reusable secmon@43000000
[    0.000000] OF: reserved mem: 0x0000000047c80000..0x0000000047d7ffff (1024 KiB) nomap non-reusable wmcpu-reserved@47c80000
[    0.000000] OF: reserved mem: 0x0000000047d80000..0x0000000047dbffff (256 KiB) nomap non-reusable wo-emi@47d80000
[    0.000000] OF: reserved mem: 0x0000000047dc0000..0x0000000047ffffff (2304 KiB) nomap non-reusable wo-data@47dc0000
[    0.000000] Zone ranges:
[    0.000000]   DMA      [mem 0x0000000040000000-0x000000004fffffff]
[    0.000000]   DMA32    empty
[    0.000000]   Normal   empty
[    0.000000] Movable zone start for each node
[    0.000000] Early memory node ranges
[    0.000000]   node   0: [mem 0x0000000040000000-0x0000000042ffffff]
[    0.000000]   node   0: [mem 0x0000000043000000-0x000000004302ffff]
[    0.000000]   node   0: [mem 0x0000000043030000-0x0000000047c7ffff]
[    0.000000]   node   0: [mem 0x0000000047c80000-0x0000000047ffffff]
[    0.000000]   node   0: [mem 0x0000000048000000-0x000000004fffffff]
[    0.000000] Initmem setup node 0 [mem 0x0000000040000000-0x000000004fffffff]
[    0.000000] psci: probing for conduit method from DT.
[    0.000000] psci: PSCIv1.1 detected in firmware.
[    0.000000] psci: Using standard PSCI v0.2 function IDs
[    0.000000] psci: MIGRATE_INFO_TYPE not supported.
[    0.000000] psci: SMC Calling Convention v1.4
[    0.000000] percpu: Embedded 18 pages/cpu s35048 r8192 d30488 u73728
[    0.000000] pcpu-alloc: s35048 r8192 d30488 u73728 alloc=18*4096
[    0.000000] pcpu-alloc: [0] 0 [0] 1
[    0.000000] Detected VIPT I-cache on CPU0
[    0.000000] CPU features: detected: GIC system register CPU interface
[    0.000000] CPU features: kernel page table isolation disabled by kernel configuration
[    0.000000] alternatives: applying boot alternatives
[    0.000000] Kernel command line:
[    0.000000] Dentry cache hash table entries: 32768 (order: 6, 262144 bytes, linear)
[    0.000000] Inode-cache hash table entries: 16384 (order: 5, 131072 bytes, linear)
[    0.000000] Built 1 zonelists, mobility grouping on.  Total pages: 64512
[    0.000000] mem auto-init: stack:off, heap alloc:off, heap free:off
[    0.000000] software IO TLB: SWIOTLB bounce buffer size adjusted to 0MB
[    0.000000] software IO TLB: area num 2.
[    0.000000] software IO TLB: SWIOTLB bounce buffer size roundup to 0MB
[    0.000000] software IO TLB: mapped [mem 0x000000004fe46000-0x000000004fec6000] (0MB)
[    0.000000] Memory: 239240K/262144K available (8768K kernel code, 992K rwdata, 2556K rodata, 448K init, 291K bss, 22904K reserved, 0K cma-reserved)
[    0.000000] SLUB: HWalign=64, Order=0-3, MinObjects=0, CPUs=2, Nodes=1
[    0.000000] rcu: Hierarchical RCU implementation.
[    0.000000] rcu:     RCU restricting CPUs from NR_CPUS=4 to nr_cpu_ids=2.
[    0.000000]  Tracing variant of Tasks RCU enabled.
[    0.000000] rcu: RCU calculated value of scheduler-enlistment delay is 10 jiffies.
[    0.000000] rcu: Adjusting geometry for rcu_fanout_leaf=16, nr_cpu_ids=2
[    0.000000] NR_IRQS: 64, nr_irqs: 64, preallocated irqs: 0
[    0.000000] GICv3: GIC: Using split EOI/Deactivate mode
[    0.000000] GICv3: 640 SPIs implemented
[    0.000000] GICv3: 0 Extended SPIs implemented
[    0.000000] Root IRQ handler: gic_handle_irq
[    0.000000] GICv3: GICv3 features: 16 PPIs
[    0.000000] GICv3: CPU0: found redistributor 0 region 0:0x000000000c080000
[    0.000000] rcu: srcu_init: Setting srcu_struct sizes based on contention.
[    0.000000] arch_timer: cp15 timer(s) running at 13.00MHz (phys).
[    0.000000] clocksource: arch_sys_counter: mask: 0xffffffffffffff max_cycles: 0x2ff89eacb, max_idle_ns: 440795202429 ns
[    0.000000] sched_clock: 56 bits at 13MHz, resolution 76ns, wraps every 4398046511101ns
[    0.000077] Calibrating delay loop (skipped), value calculated using timer frequency.. 26.00 BogoMIPS (lpj=130000)
[    0.000085] pid_max: default: 32768 minimum: 301
[    0.002997] Mount-cache hash table entries: 512 (order: 0, 4096 bytes, linear)
[    0.003005] Mountpoint-cache hash table entries: 512 (order: 0, 4096 bytes, linear)
[    0.005148] cacheinfo: Unable to detect cache hierarchy for CPU 0
[    0.005684] RCU Tasks Trace: Setting shift to 1 and lim to 1 rcu_task_cb_adjust=1 rcu_task_cpu_ids=2.
[    0.005837] rcu: Hierarchical SRCU implementation.
[    0.005840] rcu:     Max phase no-delay instances is 1000.
[    0.006252] smp: Bringing up secondary CPUs ...
[    0.006619] Detected VIPT I-cache on CPU1
[    0.006663] GICv3: CPU1: found redistributor 1 region 0:0x000000000c0a0000
[    0.006692] CPU1: Booted secondary processor 0x0000000001 [0x410fd034]
[    0.006761] smp: Brought up 1 node, 2 CPUs
[    0.006767] SMP: Total of 2 processors activated.
[    0.006770] CPU features: detected: 32-bit EL0 Support
[    0.006774] CPU features: detected: CRC32 instructions
[    0.006805] CPU features: emulated: Privileged Access Never (PAN) using TTBR0_EL1 switching
[    0.006808] CPU: All CPU(s) started at EL2
[    0.006809] alternatives: applying system-wide alternatives
[    0.010428] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 19112604462750000 ns
[    0.010445] futex hash table entries: 512 (order: 3, 32768 bytes, linear)
[    0.011706] pinctrl core: initialized pinctrl subsystem
[    0.012753] NET: Registered PF_NETLINK/PF_ROUTE protocol family
[    0.013343] DMA: preallocated 128 KiB GFP_KERNEL pool for atomic allocations
[    0.013369] DMA: preallocated 128 KiB GFP_KERNEL|GFP_DMA pool for atomic allocations
[    0.013391] DMA: preallocated 128 KiB GFP_KERNEL|GFP_DMA32 pool for atomic allocations
[    0.013766] thermal_sys: Registered thermal governor 'fair_share'
[    0.013771] thermal_sys: Registered thermal governor 'bang_bang'
[    0.013773] thermal_sys: Registered thermal governor 'step_wise'
[    0.013776] thermal_sys: Registered thermal governor 'user_space'
[    0.013828] ASID allocator initialised with 65536 entries
[    0.014547] pstore: Using crash dump compression: deflate
[    0.014552] pstore: Registered ramoops as persistent store backend
[    0.014555] ramoops: using 0x10000@0x42ff0000, ecc: 0
[    0.016005] /soc/interrupt-controller@c000000: Fixed dependency cycle(s) with /soc/interrupt-controller@c000000
[    0.021387] Modules: 29456 pages in range for non-PLT usage
[    0.021396] Modules: 520976 pages in range for PLT usage
[    0.022361] cryptd: max_cpu_qlen set to 1000
[    0.024443] SCSI subsystem initialized
[    0.024866] libata version 3.00 loaded.
[    0.026312] clocksource: Switched to clocksource arch_sys_counter
[    0.028702] NET: Registered PF_INET protocol family
[    0.028802] IP idents hash table entries: 4096 (order: 3, 32768 bytes, linear)
[    0.029854] tcp_listen_portaddr_hash hash table entries: 256 (order: 0, 4096 bytes, linear)
[    0.029866] Table-perturb hash table entries: 65536 (order: 6, 262144 bytes, linear)
[    0.029875] TCP established hash table entries: 2048 (order: 2, 16384 bytes, linear)
[    0.029893] TCP bind hash table entries: 2048 (order: 4, 65536 bytes, linear)
[    0.029946] TCP: Hash tables configured (established 2048 bind 2048)
[    0.030355] MPTCP token hash table entries: 256 (order: 0, 6144 bytes, linear)
[    0.030466] UDP hash table entries: 256 (order: 1, 8192 bytes, linear)
[    0.030482] UDP-Lite hash table entries: 256 (order: 1, 8192 bytes, linear)
[    0.030698] NET: Registered PF_UNIX/PF_LOCAL protocol family
[    0.030723] PCI: CLS 0 bytes, default 64
[    0.032013] workingset: timestamp_bits=46 max_order=16 bucket_order=0
[    0.036864] squashfs: version 4.0 (2009/01/31) Phillip Lougher
[    0.036874] jffs2: version 2.2 (NAND) (SUMMARY) (LZMA) (RTIME) (CMODE_PRIORITY) (c) 2001-2006 Red Hat, Inc.
[    0.080236] Serial: 8250/16550 driver, 3 ports, IRQ sharing disabled
[    0.081465] printk: console [ttyS0] disabled
[    0.101831] 11002000.serial: ttyS0 at MMIO 0x11002000 (irq = 72, base_baud = 2500000) is a ST16650V2
[    0.101872] printk: console [ttyS0] enabled
[    0.855967] loop: module loaded
[    0.862109] spi-nand spi0.0: calibration result: 0x3
[    0.867192] spi-nand spi0.0: ESMT SPI NAND was found.
[    0.872236] spi-nand spi0.0: 128 MiB, block size: 128 KiB, page size: 2048, OOB size: 64
[    0.880731] Signature found at block 1023 [0x07fe0000]
[    0.885864] NMBM management region starts at block 960 [0x07800000]
[    0.893094] First info table with writecount 0 found in block 960
[    0.902007] Second info table with writecount 0 found in block 963
[    0.908197] NMBM has been successfully attached
[    0.913046] 6 fixed-partitions partitions found on MTD device spi0.0
[    0.919659] Creating 6 MTD partitions on "spi0.0":
[    0.924447] 0x000000000000-0x000000100000 : "BL2"
[    0.930097] 0x000000100000-0x000000180000 : "u-boot-env"
[    0.936122] 0x000000180000-0x000000380000 : "Factory"
[    0.942865] 0x000000380000-0x0000003c0000 : "bdinfo"
[    0.948698] 0x0000003c0000-0x0000005c0000 : "FIP"
[    0.954951] 0x0000005c0000-0x0000045c0000 : "ubi"
[    0.990053] ubi0: default fastmap pool size: 25
[    0.994587] ubi0: default fastmap WL pool size: 12
[    0.999395] ubi0: attaching mtd5
[    1.220200] ubi0: scanning is finished
[    1.229752] ubi0: attached mtd5 (name "ubi", size 64 MiB)
[    1.235160] ubi0: PEB size: 131072 bytes (128 KiB), LEB size: 126976 bytes
[    1.242038] ubi0: min./max. I/O unit sizes: 2048/2048, sub-page size 2048
[    1.248822] ubi0: VID header offset: 2048 (aligned 2048), data offset: 4096
[    1.255770] ubi0: good PEBs: 512, bad PEBs: 0, corrupted PEBs: 0
[    1.261767] ubi0: user volume: 3, internal volumes: 1, max. volumes count: 128
[    1.268978] ubi0: max/mean erase counter: 11/7, WL threshold: 4096, image sequence number: 2001910725
[    1.278184] ubi0: available PEBs: 0, total reserved PEBs: 512, PEBs reserved for bad PEB handling: 19
[    1.287398] ubi0: background thread "ubi_bgt0d" started, PID 181
[    1.294093] block ubiblock0_1: created from ubi0:1(rootfs)
[    1.299625] ubiblock: device ubiblock0_1 (rootfs) set to be root filesystem
[    1.809583] mtk_soc_eth 15100000.ethernet eth0: mediatek frame engine at 0xffffffc081580000, irq 75
[    1.819477] mtk_soc_eth 15100000.ethernet eth1: mediatek frame engine at 0xffffffc081580000, irq 75
[    1.829218] i2c_dev: i2c /dev entries driver
[    1.835260] mtk-wdt 1001c000.watchdog: Watchdog enabled (timeout=31 sec, nowayout=0)
[    1.844410] NET: Registered PF_INET6 protocol family
[    1.850689] Segment Routing with IPv6
[    1.854388] In-situ OAM (IOAM) with IPv6
[    1.858389] NET: Registered PF_PACKET protocol family
[    1.863606] 8021q: 802.1Q VLAN Support v1.8
[    1.966098] mt7530-mdio mdio-bus:1f: configuring for fixed/2500base-x link mode
[    1.975158] mt7530-mdio mdio-bus:1f: Link is Up - 2.5Gbps/Full - flow control rx/tx
[    1.985898] mt7530-mdio mdio-bus:1f lan1 (uninitialized): PHY [mt7530-0:00] driver [MediaTek MT7531 PHY] (irq=79)
[    2.008198] mt7530-mdio mdio-bus:1f lan2 (uninitialized): PHY [mt7530-0:01] driver [MediaTek MT7531 PHY] (irq=80)
[    2.030225] mt7530-mdio mdio-bus:1f lan3 (uninitialized): PHY [mt7530-0:02] driver [MediaTek MT7531 PHY] (irq=81)
[    2.052302] mt7530-mdio mdio-bus:1f lan4 (uninitialized): PHY [mt7530-0:03] driver [MediaTek MT7531 PHY] (irq=82)
[    2.063876] mtk_soc_eth 15100000.ethernet eth0: entered promiscuous mode
[    2.070664] DSA: tree 0 setup
[    2.074095] clk: Disabling unused clocks
[    2.083236] VFS: Mounted root (squashfs filesystem) readonly on device 254:0.
[    2.090575] Freeing unused kernel memory: 448K
[    2.095081] Run /sbin/init as init process
[    2.099181]   with arguments:
[    2.102136]     /sbin/init
[    2.104831]   with environment:
[    2.107974]     HOME=/
[    2.110322]     TERM=linux
[    2.301218] init: Console is alive
[    2.304753] init: - watchdog -
[    2.734674] kmodloader: loading kernel modules from /etc/modules-boot.d/*
[    2.751715] gpio_button_hotplug: loading out-of-tree module taints kernel.
[    2.762986] kmodloader: done loading kernel modules from /etc/modules-boot.d/*
[    2.778978] init: - preinit -
[    3.146850] mtk_soc_eth 15100000.ethernet wan: renamed from eth1
[    3.208031] mtk_soc_eth 15100000.ethernet eth0: configuring for fixed/2500base-x link mode
[    3.218019] mtk_soc_eth 15100000.ethernet eth0: Link is Up - 2.5Gbps/Full - flow control rx/tx
[    3.245191] mt7530-mdio mdio-bus:1f lan1: configuring for phy/gmii link mode
[    5.376329] random: crng init done
[    7.487933] UBIFS (ubi0:2): Mounting in unauthenticated mode
[    7.493674] UBIFS (ubi0:2): background thread "ubifs_bgt0_2" started, PID 678
[    7.526380] UBIFS (ubi0:2): recovery needed
[    7.615776] UBIFS (ubi0:2): recovery completed
[    7.620300] UBIFS (ubi0:2): UBIFS: mounted UBI device 0, volume 2, name "rootfs_data"
[    7.628133] UBIFS (ubi0:2): LEB size: 126976 bytes (124 KiB), min./max. I/O unit sizes: 2048 bytes/2048 bytes
[    7.638036] UBIFS (ubi0:2): FS size: 42283008 bytes (40 MiB, 333 LEBs), max 343 LEBs, journal size 2158592 bytes (2 MiB, 17 LEBs)
[    7.649676] UBIFS (ubi0:2): reserved for root: 1997130 bytes (1950 KiB)
[    7.656276] UBIFS (ubi0:2): media format: w5/r0 (latest is w5/r0), UUID D2F60841-1F50-4DA5-9E4F-0849F3707601, small LPT model
[    7.671990] mount_root: switching to ubifs overlay
[    7.682717] overlayfs: null uuid detected in lower fs '/', falling back to xino=off,index=off,nfs_export=off.
[    7.698021] urandom-seed: Seeding with /etc/urandom.seed
[    7.767569] procd: - early -
[    7.770512] procd: - watchdog -
[    8.317096] procd: - watchdog -
[    8.321884] procd: - ubus -
[    8.397020] procd: - init -
[    8.770671] kmodloader: loading kernel modules from /etc/modules.d/*
[    8.822533] crypto-safexcel 10320000.crypto: EIP97:230(0,1,4,4)-HIA:270(0,5,5),PE:150/433(alg:7fcdfc00)/0/0/0
[    8.841586] Loading modules backported from Linux version v6.12.6-0-ge9d65b48ce1a
[    8.849091] Backport generated by backports.git v6.1.110-1-35-g410656ef04d2
[    8.873125] xt_time: kernel timezone is -0000
[    8.972643] urngd: v1.0.2 started.
[    9.241160] mt798x-wmac 18000000.wifi: HW/SW Version: 0x8a108a10, Build Time: 20240823161240a
[    9.241160]
[    9.442527] mt798x-wmac 18000000.wifi: WM Firmware Version: ____000000, Build Time: 20240823161304
[    9.536870] mt798x-wmac 18000000.wifi: WA Firmware Version: DEV_000000, Build Time: 20240823161841
[    9.634714] mt798x-wmac 18000000.wifi: registering led 'mt76-phy0'
[    9.717985] mt798x-wmac 18000000.wifi: registering led 'mt76-phy1'
[    9.807137] PPP generic driver version 2.4.2
[    9.812686] NET: Registered PF_PPPOX protocol family
[    9.819843] wireguard: WireGuard 1.0.0 loaded. See www.wireguard.com for information.
[    9.827692] wireguard: Copyright (C) 2015-2019 Jason A. Donenfeld <Jason@zx2c4.com>. All Rights Reserved.
[    9.847550] mtdblock: MTD device 'bdinfo' is NAND, please consider using UBI block devices instead.
[    9.858060] kmodloader: done loading kernel modules from /etc/modules.d/*
[   10.095399] mtdblock: MTD device 'bdinfo' is NAND, please consider using UBI block devices instead.
[   10.698227] mtdblock: MTD device 'ubi' is NAND, please consider using UBI block devices instead.
[   10.734644] mtdblock: MTD device 'FIP' is NAND, please consider using UBI block devices instead.
[   10.758508] mtdblock: MTD device 'bdinfo' is NAND, please consider using UBI block devices instead.
[   10.775426] mtdblock: MTD device 'Factory' is NAND, please consider using UBI block devices instead.
[   10.799313] mtdblock: MTD device 'u-boot-env' is NAND, please consider using UBI block devices instead.
[   10.816740] mtdblock: MTD device 'BL2' is NAND, please consider using UBI block devices instead.
[   13.570472] mtk_soc_eth 15100000.ethernet eth0: Link is Down
[   13.594831] mtk_soc_eth 15100000.ethernet eth0: configuring for fixed/2500base-x link mode
[   13.603311] mtk_soc_eth 15100000.ethernet eth0: Link is Up - 2.5Gbps/Full - flow control rx/tx
[   13.610075] mt7530-mdio mdio-bus:1f lan1: configuring for phy/gmii link mode
[   13.620124] br-lan: port 1(lan1) entered blocking state
[   13.625359] br-lan: port 1(lan1) entered disabled state
[   13.630665] mt7530-mdio mdio-bus:1f lan1: entered allmulticast mode
[   13.636987] mtk_soc_eth 15100000.ethernet eth0: entered allmulticast mode
[   13.651439] mt7530-mdio mdio-bus:1f lan1: entered promiscuous mode
[   13.681609] mt7530-mdio mdio-bus:1f lan2: configuring for phy/gmii link mode
[   13.689849] br-lan: port 2(lan2) entered blocking state
[   13.695084] br-lan: port 2(lan2) entered disabled state
[   13.700408] mt7530-mdio mdio-bus:1f lan2: entered allmulticast mode
[   13.710229] mt7530-mdio mdio-bus:1f lan2: entered promiscuous mode
[   13.725679] mt7530-mdio mdio-bus:1f lan3: configuring for phy/gmii link mode
[   13.737742] br-lan: port 3(lan3) entered blocking state
[   13.743000] br-lan: port 3(lan3) entered disabled state
[   13.748380] mt7530-mdio mdio-bus:1f lan3: entered allmulticast mode
[   13.759697] mt7530-mdio mdio-bus:1f lan3: entered promiscuous mode
[   13.776228] mt7530-mdio mdio-bus:1f lan4: configuring for phy/gmii link mode
[   13.787668] br-lan: port 4(lan4) entered blocking state
[   13.792918] br-lan: port 4(lan4) entered disabled state
[   13.798232] mt7530-mdio mdio-bus:1f lan4: entered allmulticast mode
[   13.806989] mt7530-mdio mdio-bus:1f lan4: entered promiscuous mode
[   14.056581] mtk_soc_eth 15100000.ethernet wan: PHY [mdio-bus:06] driver [RTL8221B-VB-CG 2.5Gbps PHY (C45)] (irq=POLL)
[   14.067236] mtk_soc_eth 15100000.ethernet wan: configuring for phy/2500base-x link mode
[   17.010176] br-lan: port 5(phy0-ap0) entered blocking state
[   17.015757] br-lan: port 5(phy0-ap0) entered disabled state
[   17.021447] mt798x-wmac 18000000.wifi phy0-ap0: entered allmulticast mode
[   17.028485] mt798x-wmac 18000000.wifi phy0-ap0: entered promiscuous mode
[   17.035315] br-lan: port 5(phy0-ap0) entered blocking state
[   17.040918] br-lan: port 5(phy0-ap0) entered forwarding state
[   17.047201] br-lan: port 5(phy0-ap0) entered disabled state
[   17.098784] mt7530-mdio mdio-bus:1f lan4: Link is Up - 1Gbps/Full - flow control off
[   17.108223] br-lan: port 4(lan4) entered blocking state
[   17.113461] br-lan: port 4(lan4) entered forwarding state
[   17.432873] br-lan: port 6(phy1-ap0) entered blocking state
[   17.438538] br-lan: port 6(phy1-ap0) entered disabled state
[   17.444247] mt798x-wmac 18000000.wifi phy1-ap0: entered allmulticast mode
[   17.451319] mt798x-wmac 18000000.wifi phy1-ap0: entered promiscuous mode
[   17.458332] br-lan: port 6(phy1-ap0) entered blocking state
[   17.463937] br-lan: port 6(phy1-ap0) entered forwarding state
[   17.989270] br-lan: port 7(phy1-mesh0) entered blocking state
[   17.995055] br-lan: port 7(phy1-mesh0) entered disabled state
[   18.000881] mt798x-wmac 18000000.wifi phy1-mesh0: entered allmulticast mode
[   18.008125] mt798x-wmac 18000000.wifi phy1-mesh0: entered promiscuous mode
[   19.116483] br-lan: port 6(phy1-ap0) entered disabled state
[   19.813976] br-lan: port 5(phy0-ap0) entered blocking state
[   19.819574] br-lan: port 5(phy0-ap0) entered forwarding state
[   20.181293] br-lan: port 7(phy1-mesh0) entered blocking state
[   20.187078] br-lan: port 7(phy1-mesh0) entered forwarding state
[   20.200217] br-lan: port 6(phy1-ap0) entered blocking state
[   20.205823] br-lan: port 6(phy1-ap0) entered forwarding state
```

Once started in order to get internet again I apply the workaround
`[   55.836986] mtk_soc_eth 15100000.ethernet wan: Link is Up - 1Gbps/Full - flow control off`

aiamadeus commented:
> AFAIK with my tests this issue is related to a bad PHY register/reference in dts
> In this proposal the reg is OK and the the reference is ok , but with C45 driver I never got link up , please keep c22

Keep c22 for now so this pr won't break it.

aiamadeus commented:
> AFAIK with my tests this issue is related to a bad PHY register/reference in dts
> In this proposal the reg is OK and the the reference is ok , but with C45 driver I never got link up , please keep c22

Changing tr3000 to c22 didn't fix phy init, so I guess it's not the same issue.
Keep c22 for wr3000h so this pr won't break it.

dangowrt commented:
Possible fix for problems with RTL8221B: https://github.com/openwrt/openwrt/pull/17701

juampe commented:
> Sorry, but c22 with resets and without phy-mode is not tested
> Yesterday I tested your patch, c45 with resets and without phy-mode with ethtool workaround with Ok result
> I'll do some test today with main

@aimadeus The actual state of the commit is not operational without workaround

The weird thing is that wan is detected as C45 (I built it twice with clean to make sure).

```
	phy6: ethernet-phy@6 {
		compatible = "ethernet-phy-ieee802.3-c22";
		reg = <6>;
		reset-assert-us = <100000>;
		reset-deassert-us = <100000>;
		reset-gpios = <&pio 3 GPIO_ACTIVE_LOW>;
	};
```

```
root@wrt2:~# dmesg|grep wan
[    3.160594] mtk_soc_eth 15100000.ethernet wan: renamed from eth1
[   14.766538] mtk_soc_eth 15100000.ethernet wan: PHY [mdio-bus:06] driver [RTL8221B-VB-CG 2.5Gbps PHY (C45)] (irq=POLL)
[   14.777178] mtk_soc_eth 15100000.ethernet wan: configuring for phy/2500base-x link mode
```

```
root@wrt2:~# /usr/sbin/ethtool -s wan autoneg on speed 1000 duplex full
root@wrt2:~# dmesg|grep wan
[    3.160594] mtk_soc_eth 15100000.ethernet wan: renamed from eth1
[   14.766538] mtk_soc_eth 15100000.ethernet wan: PHY [mdio-bus:06] driver [RTL8221B-VB-CG 2.5Gbps PHY (C45)] (irq=POLL)
[   14.777178] mtk_soc_eth 15100000.ethernet wan: configuring for phy/2500base-x link mode
[  125.197036] mtk_soc_eth 15100000.ethernet wan: Link is Up - 1Gbps/Full - flow control off


```

aiamadeus commented:
However, defining phy mode on phy node is useless,
it should be on mac node (via of_get_phy_mode).
I think the difference is whether to use gpio-export or standard Ethernet PHY bindings to reset the phy.

aiamadeus commented:
Btw c22 (of rtl8221) requires mac to work into fixed-link mode. Usually it should be c45 mode.
